katsdpcontroller/product_config.py

```diff
@@ -31,63 +82,1196 @@ class _Sensor(ABC):
+            archive=config['archive'],
+            n_servers=4 if not options.develop else 2
+        )
+
+    def compatible(self, other: 'VisStream') -> bool:
+        """Determine whether the configurations are mostly the same.
+
+        Specifically, they must be the same other than the vlaues of
+        ``continuum_factor`` and ``archive``.
+        """
+        return (
+            self.src_streams[0] is other.src_streams[0]
+            and self.int_time == other.int_time
+            and self.output_channels == other.output_channels
+            and self.excise == other.excise
+            and self.n_servers == other.n_servers
+        )
+
+
+class BeamformerStreamBase(Stream):
+    """Base for sdp.beamformer and sdp.beamformer_engineering streams."""
+
+    _valid_src_types: ClassVar[_ValidTypes] = {
+        'cbf.tied_array_channelised_voltage',
+        'sim.cbf.tied_array_channelised_voltage'
+    }
+
+    def __init__(self, name: str, src_streams: Sequence[Stream]) -> None:
+        super().__init__(name, src_streams)
+        acv = self.antenna_channelised_voltage
+        if not all(stream.src_streams[0] is acv for stream in src_streams):
+            raise ValueError('Source streams do not come from the same channeliser')
+
+    @property
+    def antenna_channelised_voltage(self) -> AntennaChannelisedVoltageStreamBase:
+        return cast(AntennaChannelisedVoltageStreamBase, self.src_streams[0].src_streams[0])
+
+    @property
+    def tied_array_channelised_voltage(self) -> Sequence[TiedArrayChannelisedVoltageStreamBase]:
+        return [
+            cast(TiedArrayChannelisedVoltageStreamBase, stream)
+            for stream in self.src_streams
+        ]
+
+    @property
+    def n_channels(self) -> int:
+        return self.antenna_channelised_voltage.n_channels
+
+
+class BeamformerStream(BeamformerStreamBase):
+    """Instance of sdp.beamformer."""
+
+    stream_type = 'sdp.beamformer'
+
+    @classmethod
+    def from_config(cls,
+                    options: Options,
+                    name: str,
+                    config: Mapping[str, Any],
+                    src_streams: Sequence['Stream'],
+                    sensors: Mapping[str, Any]) -> 'BeamformerStream':
+        return cls(name, src_streams)
+
+
+class BeamformerEngineeringStream(BeamformerStreamBase):
+    """Instance of sdp.beamformer_engineering."""
+
+    stream_type = 'sdp.beamformer_engineering'
+
+    def __init__(self, name: str, src_streams: Sequence[Stream], *,
+                 store: str,
+                 output_channels: Optional[Tuple[int, int]] = None) -> None:
+        super().__init__(name, src_streams)
+        cbf_channels = self.antenna_channelised_voltage.n_channels
+        c = _normalise_output_channels(cbf_channels, output_channels)
+        for tacv in self.tied_array_channelised_voltage:
+            c = _normalise_output_channels(cbf_channels, c, tacv.n_channels_per_endpoint)
```

Comment:
Not sure why these need to be normalised per endpoint, but will ask you over video about this.

Reply:
For the very hypothetical case that the two streams have different numbers of endpoints, you'd want to align the output channel range to the coarser of the two (actually it would need to be aligned to the LCM of the two, but as long as we're dealing with power-of-two channelisation the larger value will always be a multiple of the smaller one).